src/agent/src/mount.rs

```diff
@@ -51,6 +52,75 @@ const RO_MASK: u32 = 0o440;
 const EXEC_MASK: u32 = 0o110;
 const MODE_SETGID: u32 = 0o2000;
 
+pub trait RefCounted {
+    // Increment the reference count by one
+    fn inc_ref(&self);
+
+    // Decrement the reference count by one
+    fn dec_ref(&self) -> Result<()>;
+
+    // Return true if the reference count is zero
+    fn is_zero(&self) -> bool;
+
+    //Return the count number
+    fn get_count(&self) -> u32;
+}
+#[derive(Debug)]
+pub struct RefCountedObject<T> {
+    pub name: String,
+    pub ref_count: AtomicU32,
+    pub ref_type: T,
+}
+
+// Implement the RefCounted trait for the generic struct
+impl<T> RefCounted for RefCountedObject<T> {
+    fn inc_ref(&self) {
+        self.ref_count.fetch_add(1, Ordering::Relaxed);
+    }
+
+    fn dec_ref(&self) -> Result<()> {
+        self.ref_count.fetch_sub(1, Ordering::Relaxed);
+        Ok(())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.ref_count.load(Ordering::Relaxed) == 0
+    }
+
+    fn get_count(&self) -> u32 {
+        self.ref_count.load(Ordering::Relaxed)
+    }
+}
+#[derive(Debug)]
+// Define an enum to represent the MountPoint and VerityDevice variants
+pub enum ObjectKind {
+    MountPoint,
+    VerityDevice,
+}
+
+// Implement a custom dec_ref method for each variant
+impl RefCountedObject<ObjectKind> {
+    fn dec_ref(&self) -> Result<()> {
+        self.ref_count.fetch_sub(1, Ordering::Relaxed);
```

Comment:
If this is truly used concurrently, this should be a release operation.

And we should have an acquire fence after checking that the refcount is zero so that we can ensure that all usage of the object happens-before it is torn down.